### PR TITLE
feat: creates a card pile

### PR DESCRIPTION
**Pull request by guilhermevrs**
related #25 

So here is my line of thoughts:

- The logic of shuffling / cutting / getting a card should be in this component
- The style effects on how to differentiate a single card from a card pile should be in this component
- I don't want to generate the cards inside the card pile, otherwise I would need a pile for each type of card...

Then it hit me: Apart from shuffling, any other operation should return card(s) and possibly need to destroy the own component (e.g. when cutting, the card pile should be destroyed and two other piles should appear).

What I would like to build is a component that receives the cards via children (so any card can be inserted) and have like a context menu where you can choose what to do (e.g. shuffle, cut, and so on).

For getting a card (no matter if its top, bottom, middle, random, whatever), this component would have an event like onCardRequested.
For cutting, this component would have an event like onCut
And when the action performed means to destroy the pile (when cutting, when removing the last card), then the component would fire another event like onDestroy. Another option would be not have the onDestroy, but on the above events have an event options saying that the pile should not live anymore.
Someone else listening to this (could be the table) would then manage the display and rearrange.

Now for the problem: Since I want to only control what the card does via a context menu, I should have something like flip the first card. This would then, well, flip the first card :)
But:

- If I want to simply open the context menu on click, means that I should not listen anymore for click events when in a card pile (managed by disableNativeEvents prop)
- And when the card pile decides to flip the first card, then something needs to happen:
     - I tried to use the faceUp prop, but this re render the card and I don't have the nice animation as before
     - I could forwardRef the card and use the useImperativeHandle, but is this the React way of doing it? I tried already and it works, btw.

So here is my work so far :)

### CODE REVIEW

**Comment by tib-tib** [comment]
I am not sure about destroying the pile when shuffling or cutting it. After all, it's all about reordering its children, so it should only be a rerender.

Regarding passing the cards through children, I don't really like it (mostly because of: ```const childrenList = React.Children.toArray(children).filter<ReactElement<CardProps>>(React.isValidElement);```) 

I think it could be something like this instead:

```
<CardPile cards={cards} component={PlayingCard} />
```

But I agree it is something we need to be careful about if we want it to be generic.

Do you want to discuss it more in a call? It would probably be more efficient for a brainstorm 😄 

Anyway, nice work so far 💪

File: src/utils/array-utils.tsx [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    randomIndex = Math.floor(Math.random() * currentIndex);
+    currentIndex -= 1;
+
+    // And swap it with the current element.
+    temporaryValue = newPile[currentIndex];
+    newPile[currentIndex] = newPile[randomIndex];
+    newPile[randomIndex] = temporaryValue;
+  }
+  return newPile;
+}
```
Comment (tib-tib) [inline src/utils/array-utils.tsx:23]: What about something simpler like:

```
export function shuffleArray<T = unknown>(cards: readonly T[]): T[] {
  return cards.slice().sort(() => Math.random() - 0.5);
}
```

Comment (guilhermevrs) [reply src/utils/array-utils.tsx:23]: Indeed it's a good idea to simplify it
I don't know if the probabilities that a card stays at the same place are acceptable

Let's go with that that indeed simplifies a lot and we will what's going to happen later

File: src/components/molecules/CardPile/CardPile.tsx [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+  },
+  cardPileMenu: {
+    position: 'absolute',
+    left: '-100%',
+  },
+});
+
+// eslint-disable-next-line import/prefer-default-export
+export const CardPile: FC<CardPileProps> = ({ cards, component = PlayingCard, backFace = PlayingCardFace }) => {
+  const { cardPile, multiItem, card, firstItem, secondItem, lastItem, overlay, cardPileMenu } = useStyles();
```
Comment (tib-tib) [inline src/components/molecules/CardPile/CardPile.tsx:82]: I don't really like using destructuring for class names. It looks like we are using a cardPile object, a card, etc. Why don't you do `const classes = useStyles();` and then use them like `className={classes.cardPile}`?

Comment (guilhermevrs) [reply src/components/molecules/CardPile/CardPile.tsx:82]: I don't like it either!
I had the impression it was my linter complaining, but perhaps it was on another case.
I will move it back to classes

**Review by tib-tib — APPROVED** [review]

**Pull request merged**